sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala

```diff
@@ -81,6 +93,147 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
+          case None => (nodeUpdateInfo.history, Success(nodeUpdateInfo.state), nodeUpdateInfo.wallet, nodeUpdateInfo.suffix)
+        }
+      case (Failure(e), _) =>
+        log.error("State rollback failed: ", e)
+        context.system.eventStream.publish(RollbackFailed)
+        //todo: what to return here? the situation is totally wrong
+        ???
+      case (_, Failure(e)) =>
+        log.error("Wallet rollback failed: ", e)
+        context.system.eventStream.publish(RollbackFailed)
+        //todo: what to return here? the situation is totally wrong
+        ???
+    }
+  }
+
+  // This method is actually a copy-paste of parent NodeViewHolder.trimChainSuffix method.
+  protected def trimChainSuffix(suffix: IndexedSeq[SidechainBlock], rollbackPoint: scorex.util.ModifierId): IndexedSeq[SidechainBlock] = {
+    val idx = suffix.indexWhere(_.id == rollbackPoint)
+    if (idx == -1) IndexedSeq() else suffix.drop(idx)
+  }
+
+  // Apply state and wallet with blocks one by one, if consensus epoch is going to be changed -> notify wallet and history.
+  protected def applyStateAndWallet(history: HIS,
+                           stateToApply: MS,
+                           walletToApply: VL,
+                           suffixTrimmed: IndexedSeq[SidechainBlock],
+                           progressInfo: ProgressInfo[SidechainBlock]): SidechainNodeUpdateInformation = {
+    val updateInfoSample = SidechainNodeUpdateInformation(history, stateToApply, walletToApply, None, None, suffixTrimmed)
+    progressInfo.toApply.foldLeft(updateInfoSample) { case (updateInfo, modToApply) =>
+      if (updateInfo.failedMod.isEmpty) {
+        // Check if the next modifier will change Consensus Epoch, so notify History and Wallet with current info.
+        val (newHistory, newWallet) = if(updateInfo.state.isSwitchingConsensusEpoch(modToApply)) {
+          val (lastBlockInEpoch, consensusEpochInfo) = updateInfoSample.state.getCurrentConsensusEpochInfo
+          val historyAfterStakeConsensusApply = updateInfo.history.applyStakeConsensusEpochInfo(
+            lastBlockInEpoch,
+            StakeConsensusEpochInfo(consensusEpochInfo.merkleTree.rootHash(), consensusEpochInfo.totalStake)
+          )
+          val walletAfterStakeConsensusApply = updateInfoSample.wallet.applyConsensusEpochInfo(consensusEpochInfo)
+          (historyAfterStakeConsensusApply, walletAfterStakeConsensusApply)
+        } else
+          (updateInfo.history, updateInfo.wallet)
+
+        updateInfo.state.applyModifier(modToApply) match {
+          case Success(stateAfterApply) =>
+            val walletAfterApply = newWallet.scanPersistent(modToApply)
+            val historyAfterApply = newHistory.reportModifierIsValid(modToApply)
+            context.system.eventStream.publish(SemanticallySuccessfulModifier(modToApply))
+            SidechainNodeUpdateInformation(historyAfterApply, stateAfterApply, walletAfterApply, None, None, updateInfo.suffix :+ modToApply)
```

Comment:
look like we have here unnecessary copy here in "updateInfo.suffix :+ modToApply" we will do a deep copy of updateInfo.suffix plus modToApply. Shall we change suffix type from IndexedSeq to List?